### PR TITLE
Merge in null safety changes with additional fixes.

### PR DESCRIPTION
**Pull request by markegli**
This includes the internal null safety migration, as well as bug fixes for errors that were only caught when the tests were run through pub.

### CODE REVIEW

File: test/list_differ_test.dart [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import 'package:observable/observable.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('$ListDiffer', () {
```
Comment (ncroll) [inline test/list_differ_test.dart:5]: I've seen some tests just supply the type. `group(ListDiffer, () {

I don't feel strongly about it, but something to consider.

File: test/list_differ_test.dart [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    final diff = const ListDiffer<String>().diff;
```
Comment (ncroll) [inline test/list_differ_test.dart:6]: What about `const diff = ListDiffer<String>().diff`?

**Review by ncroll — APPROVED** [review]

File: test/list_differ_test.dart (hunk 1 above)
Comment (markegli) [reply test/list_differ_test.dart:5]: Ack. I'll keep it consistent for now. If I change this I'll do it as part of a larger change that also updates the other tests.

File: test/list_differ_test.dart (hunk 2 above)
Comment (markegli) [reply test/list_differ_test.dart:6]: Sadly this doesn't work because the ListDiffer object is const but its `diff` property is not.

**Pull request merged**